src/java/org/apache/cassandra/index/sai/disk/v1/TermsReader.java

```diff
@@ -203,6 +214,93 @@ public PostingsReader getPostingReader(long offset) throws IOException
         }
     }
 
+    public class RangeQuery
+    {
+        private final QueryEventListener.TrieIndexEventListener listener;
+        private final long lookupStartTime;
+        private final QueryContext context;
+
+        private final Expression exp;
+
+        RangeQuery(Expression exp, QueryEventListener.TrieIndexEventListener listener, QueryContext context)
+        {
+            this.listener = listener;
+            this.exp = exp;
+            lookupStartTime = System.nanoTime();
+            this.context = context;
+        }
+
+        public PostingList execute()
+        {
+            final ByteComparable lower = exp.lower != null ? ByteComparable.fixedLength(exp.getLowerBound()) : null;
+            final ByteComparable upper = exp.upper != null ? ByteComparable.fixedLength(exp.getUpperBound()) : null;
+            try (TrieRangeIterator reader = new TrieRangeIterator(termDictionaryFile.instantiateRebufferer(),
+                                                                  termDictionaryRoot,
+                                                                  lower,
+                                                                  upper,
+                                                                  exp.lowerInclusive,
+                                                                  exp.upperInclusive))
+            {
+                var iter = reader.iterator();
+                if (!iter.hasNext())
+                    return PostingList.EMPTY;
+
+                context.checkpoint();
+                // Because postings are not sorted, we need to eagerly materialize the results and sort them.
+                LongHeap postings = materializeResults(iter);
+
+                listener.onTraversalComplete(System.nanoTime() - lookupStartTime, TimeUnit.NANOSECONDS);
+
+                return new LongHeapPostingList(postings);
+            }
+            catch (Throwable e)
+            {
+                if (!(e instanceof AbortedOperationException))
+                    logger.error(indexContext.logMessage("Failed to execute term query"), e);
+
+                throw Throwables.cleaned(e);
+            }
+        }
+
+        /**
+         * Build an in-memory heap of row ids from the posting lists of the matching terms.
+         * @return an ordered {@link LongHeap} of row ids
+         */
+        private LongHeap materializeResults(Iterator<Pair<ByteSource,Long>> trieEntries) throws IOException
+        {
+            assert trieEntries.hasNext();
+            var heap = new LongHeap(1);
```

Comment:
Is `1` a good starting value for this heap? The lucene classes are pretty good at array resizing but `1` means growth is starting from the minimal state. One option would be to lazily instantiate the heap after the postings reader is opened and use `currentReader.size()` as the initial size.

Reply:
I wasn't really sure which size to go with. The `currentReader.size()` is just the number of rows that have a given map entry, which is likely better than a hard coded 1, but might also be just one entry in many cases.

Reply:
I went with your suggestion. However, `currentReader.size()` is a long, and the `LongHeap` takes an `int` for size. Do we actually have more than integer max number of rows in a given sstable?

Reply:
> Do we actually have more than integer max number of rows in a given sstable?

I'm pretty certain you can. It's one of the index segmentation rules to segment when we hit that constraint. Saying that this is searching on a segment so I don't think it can hit more than an integer max number of rows. But, only if you don't compact the segments.